### PR TITLE
feat(iam): local user password complexity

### PR DESCRIPTION
**Pull request by GaryJBlake**
In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/blob/main/CONTRIBUTING.md) for making a pull request.

**Summary of Pull Request**

- Added `Get-LocalPasswordComplexity` cmdlet to retrieve the password complexity in `/etc/pam.d/system-password`.
- Added `Set-LocalPasswordComplexity` cmdlet to configure the password complexity in `/etc/pam.d/system-password`.
- Added `Request-VcenterPasswordComplexity` cmdlet to retrieve the password complexity for a vCenter Server.
- Added `Update-VcenterPasswordComplexity` cmdlet to update the password complexity for a vCenter Server.
- Added `Request-NsxtManagerPasswordComplexity` cmdlet to retrieve the password complexity for NSX Manager nodes.
- Added `Update-NsxtManagerPasswordComplexity` cmdlet to update the password complexity for NSX Manager nodes.
- Added `Request-NsxtEdgePasswordComplexity` cmdlet to retrieve the password complexity for NSX Edge nodes.
- Added `Update-NsxtEdgePasswordComplexity` cmdlet to update the password complexity for NSX Edge nodes.
- Moved `Get-NsxtManagerAuthenticationPolicy` to Deprecation Region
- Moved `Set-NsxtManagerAuthenticationPolicy` to Deprecation Region
- Moved `Get-NsxtEdgeNodeAuthenticationPolicy` to Deprecation Region
- Moved `Set-NsxtEdgeNodeAuthenticationPolicy` to Deprecation Region

Signed-off-by: Gary Blake <gblake@vmware.com>

**Type of Pull Request**

- [ ] This is a bug fix.
- [ ] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.

**Related to Existing Issues**

Issue Number: N/A

**Test and Documentation Coverage**

- [ ] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [ ] No, there are no breaking changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -32,6 +32,14 @@
 - Added `Update-NsxtManagerAccountLockout` cmdlet to update the account lockout policy for NSX Manager nodes.
 - Added `Request-NsxtEdgeAccountLockout` cmdlet to retrieve the account lockout policy for NSX Edge nodes.
 - Added `Update-NsxtEdgeAccountLockout` cmdlet to update the account lockout policy for NSX Edge nodes.
+- Added `Get-LocalPasswordComplexity` cmdlet to retrieve the password complexity in /etc/pam.d/system-password.
```
Comment (tenthirtyam) [inline CHANGELOG.md:35]:
```suggestion
- Added `Get-LocalPasswordComplexity` cmdlet to retrieve the password complexity in `/etc/pam.d/system-password`.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -32,6 +32,14 @@
...
+- Added `Set-LocalPasswordComplexity` cmdlet to configure the password complexity in /etc/pam.d/system-password.
```
Comment (tenthirtyam) [inline CHANGELOG.md:36]:
```suggestion
- Added `Set-LocalPasswordComplexity` cmdlet to configure the password complexity in `/etc/pam.d/system-password`.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -32,6 +32,14 @@
...
+- Added `Request-VcenterPasswordComplexity` cmdlet to retrieve the password complexity for a vCenter Server.
+- Added `Update-VcenterPasswordComplexity` cmdlet to update the password complexity for a vCenter Server.
+- Added `Request-NsxtManagerPasswordComplexity` cmdlet to retrieve the password complexity for NSX Manager nodes.
+- Added `Update-NsxtManagerPasswordComplexity` cmdlet to update the password complexity for NSX Manager nodes.
```
Comment (tenthirtyam) [inline CHANGELOG.md:40]:
```suggestion
- Added `Update-NsxtManagerPasswordComplexity` cmdlet to update the password complexity for NSX Local Manager nodes.
```

File: PowerValidatedSolutions.psm1 [hunk 4]
```diff
@@ -16870,6 +16714,77 @@ Export-ModuleMember -Function Update-VcenterRootPasswordExpiration
...
+                                                    $NsxtManagerPasswordComplexityObject | Add-Member -notepropertyname "Min Numerical" -notepropertyvalue $nsxtManagerNodePolicy.'Min Numerical'
+                                                    $NsxtManagerPasswordComplexityObject | Add-Member -notepropertyname "Min Special" -notepropertyvalue $nsxtManagerNodePolicy.'Min Special'
+                                                    $NsxtManagerPasswordComplexityObject | Add-Member -notepropertyname "Min Unique" -notepropertyvalue $nsxtManagerNodePolicy.'Min Unique'
+                                                    $NsxtManagerPasswordComplexityObject | Add-Member -notepropertyname "Max Retries" -notepropertyvalue $nsxtManagerNodePolicy.'Max Retries'
+
+                                                    $nsxtPasswordComplexityPolicy += $NsxtManagerPasswordComplexityObject
+                                                } else {
+                                                    Write-Error "Unable to retrieve account lockout policy from NSX Manager node  ($($nsxtManagerNode.fqdn)): PRE_VALIDATION_FAILED"
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16767]:
```suggestion
                                                    Write-Error "Unable to retrieve account lockout policy from NSX Manager node ($($nsxtManagerNode.fqdn)): PRE_VALIDATION_FAILED"
```

File: CHANGELOG.md (hunk 2 above)
Comment (GaryJBlake) [reply CHANGELOG.md:36]: Fixed

File: CHANGELOG.md [hunk 5]
```diff
@@ -32,6 +32,14 @@
 - Added `Update-NsxtManagerAccountLockout` cmdlet to update the account lockout policy for NSX Manager nodes.
 - Added `Request-NsxtEdgeAccountLockout` cmdlet to retrieve the account lockout policy for NSX Edge nodes.
 - Added `Update-NsxtEdgeAccountLockout` cmdlet to update the account lockout policy for NSX Edge nodes.
+- Added `Get-LocalPasswordComplexity` cmdlet to retrieve the password complexity in /etc/pam.d/system-password.
+- Added `Set-LocalPasswordComplexity` cmdlet to configure the password complexity in /etc/pam.d/system-password.
+- Added `Request-VcenterPasswordComplexity` cmdlet to retrieve the password complexity for a vCenter Server.
+- Added `Update-VcenterPasswordComplexity` cmdlet to update the password complexity for a vCenter Server.
+- Added `Request-NsxtManagerPasswordComplexity` cmdlet to retrieve the password complexity for NSX Manager nodes.
```
[the remark this answers is not in the document]
Comment (GaryJBlake) [reply CHANGELOG.md:39]: Fixed

File: CHANGELOG.md (hunk 3 above)
Comment (GaryJBlake) [reply CHANGELOG.md:40]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 4 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:16767]: Fixed

File: PowerValidatedSolutions.psm1 [hunk 6]
```diff
@@ -16870,6 +16714,77 @@ Export-ModuleMember -Function Update-VcenterRootPasswordExpiration
...
+Function Request-NsxtManagerPasswordComplexity {
+    <#
+		.SYNOPSIS
+		Retrieve the password complexity policy
+
+        .DESCRIPTION
+        The Request-NsxtManagerPasswordComplexity cmdlet retrieves the password complexity policy for each NSX Manager
+        node for a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Manager
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16726]:
```suggestion
        - Validates that network connectivity and authentication is possible to NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 7]
```diff
@@ -16932,6 +16847,78 @@ Function Request-NsxtManagerAccountLockout {
...
+Function Update-NsxtManagerPasswordComplexity {
+    <#
+		.SYNOPSIS
+		Update the password complexity policy
+
+        .DESCRIPTION
+        The Update-NsxtManagerPasswordComplexity cmdlet updates the password complexity policy for each NSX Manager
+        node for a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Manager
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16859]:
```suggestion
        - Validates that network connectivity and authentication is possible to NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 8]
```diff
@@ -16932,6 +16847,78 @@ Function Request-NsxtManagerAccountLockout {
 }
 Export-ModuleMember -Function Request-NsxtManagerAccountLockout
 
+Function Update-NsxtManagerPasswordComplexity {
+    <#
+		.SYNOPSIS
+		Update the password complexity policy
+
+        .DESCRIPTION
+        The Update-NsxtManagerPasswordComplexity cmdlet updates the password complexity policy for each NSX Manager
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16856]:
```suggestion
        The Update-NsxtManagerPasswordComplexity cmdlet updates the password complexity policy for each NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 9]
```diff
@@ -16932,6 +16847,78 @@ Function Request-NsxtManagerAccountLockout {
...
+        node for a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Manager
+		- Updates the password complexity policy
+
+        .EXAMPLE
+        Update-NsxtManagerPasswordComplexity -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -minLength 15 -minLowercase -1 -minUppercase -1  -minNumerical -1 -minSpecial -1 -minUnique 4 -maxRetry 3
+        This example updates the password complexity policy for each NSX Manager node for a workload domain
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16864]:
```suggestion
        This example updates the password complexity policy for each NSX Local Manager node for a workload domain
```

File: PowerValidatedSolutions.psm1 [hunk 10]
```diff
@@ -17027,6 +17014,77 @@ Export-ModuleMember -Function Update-NsxtManagerAccountLockout
 ##########################################################################
 #Region     Begin NSX Edge Password Management Function             ######
 
+Function Request-NsxtManagerPasswordComplexity {
+    <#
+		.SYNOPSIS
+		Retrieve the password complexity policy
+
+        .DESCRIPTION
+        The Request-NsxtManagerPasswordComplexity cmdlet retrieves the password complexity policy for each NSX Manager
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17023]:
```suggestion
        The Request-NsxtManagerPasswordComplexity cmdlet retrieves the password complexity policy for each NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 11]
```diff
@@ -17027,6 +17014,77 @@ Export-ModuleMember -Function Update-NsxtManagerAccountLockout
...
+        node for a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Manager
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17026]:
```suggestion
        - Validates that network connectivity and authentication is possible to NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 12]
```diff
@@ -17027,6 +17014,77 @@ Export-ModuleMember -Function Update-NsxtManagerAccountLockout
...
+		- Retrieves the password complexity policy
+
+        .EXAMPLE
+        Request-NsxtManagerPasswordComplexity -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01
+        This example retrieves the password complexity policy for each NSX Manager node for a workload domain
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17031]:
```suggestion
        This example retrieves the password complexity policy for each NSX Local Manager node for a workload domain
```

File: PowerValidatedSolutions.psm1 [hunk 13]
```diff
@@ -17087,6 +17145,79 @@ Function Request-NsxtEdgeAccountLockout {
...
+                                                    $updatedConfiguration = Get-LocalPasswordComplexity -vmName $nsxtEdgeNode.display_name -guestUser $vcfNsxDetails.rootUser -guestPassword $vcfNsxDetails.rootPass -nsx
+                                                    if ($updatedConfiguration.'Min Length' -eq $minLength -and $updatedConfiguration.'Min Lowercase' -eq $minLowercase -and $updatedConfiguration.'Min Uppercase' -eq $minUppercase -and $updatedConfiguration.'Min Numerical' -eq $minNumerical -and $updatedConfiguration.'Min Special' -eq $minSpecial -and $updatedConfiguration.'Min Unique' -eq $minUnique -and $updatedConfiguration.'Max Retries' -eq $maxRetry) {
+                                                        Write-Output "Update Password Complexity Policy on NSX Manager Node ($($nsxtEdgeNode.display_name)): SUCCESSFUL"
+                                                    } else {
+                                                        Write-Error "Update Password Complexity Policy on NSX Manager Node ($($nsxtEdgeNode.display_name)): POST_VALIDATION_FAILED"
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17198]: Edge Node?

```suggestion
                                                        Write-Error "Update Password Complexity Policy on NSX Edge Node ($($nsxtEdgeNode.display_name)): POST_VALIDATION_FAILED"
```

Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17198]: Edge Node?

```suggestion
                                                        Write-Error "Update Password Complexity Policy on NSX Edge Node ($($nsxtEdgeNode.display_name)): POST_VALIDATION_FAILED"
```

File: PowerValidatedSolutions.psm1 [hunk 14]
```diff
@@ -17087,6 +17145,79 @@ Function Request-NsxtEdgeAccountLockout {
...
+                                                    }
+                                                } else {
+                                                    Write-Warning "Update Password Complexity Policy on NSX Manager Node ($($nsxtEdgeNode.display_name)), already set: SKIPPED"
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17201]: Edge Node?

```suggestion
                                                    Write-Warning "Update Password Complexity Policy on NSX Edge Node ($($nsxtEdgeNode.display_name)), already set: SKIPPED"
```

File: PowerValidatedSolutions.psm1 [hunk 15]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+			a) api_failed_auth_lockout_period (in sec)
+			b) api_failed_auth_reset_period (in sec)
+			c) api_max_auth_failures (in attempt)
+			d) cli_failed_auth_lockout_period (in sec)
+			e) cli_max_auth_failures (in attempt)
+			f) minimum_password_length (in characters)
+        
+            .EXAMPLE
+        Get-NsxtManagerAuthenticationPolicy -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01
+        This example retrieves the current Authentication policy from NSX manager nodes in sfo-m01 workload domain
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37404]:
```suggestion
        This example retrieves the current Authentication policy from NSX Local Manager nodes in sfo-m01 workload domain
```

File: PowerValidatedSolutions.psm1 [hunk 16]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
 }
 Export-ModuleMember -Function Get-SsoPasswordPolicies
 
+Function Get-NsxtManagerAuthenticationPolicy {
+    <#
+		.SYNOPSIS
+        Retrieve the current Authentication Policy from NSX Manager Nodes
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37386]:
```suggestion
        Retrieve the current Authentication Policy from NSX Local Manager Nodes
```

File: PowerValidatedSolutions.psm1 [hunk 17]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+    } Catch {
+        Debug-ExceptionWriter -object $_
+    }
+}
+Export-ModuleMember -Function Get-NsxtManagerAuthenticationPolicy
+
+Function Set-NsxtManagerAuthenticationPolicy {
+    <#
+		.SYNOPSIS
+        Configure Authentication Password Policy NSX Manager Nodes
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37440]:
```suggestion
        Configure Authentication Password Policy NSX Local Manager Nodes
```

File: PowerValidatedSolutions.psm1 [hunk 18]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+
+        .DESCRIPTION
+        The Set-NsxtManagerAuthenticationPolicy cmdlet configures Authentication policy within NSX manager nodes within
+        a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Management Cluster
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37446]:
```suggestion
        - Validates that network connectivity and authentication is possible to NSX Manager
```

File: PowerValidatedSolutions.psm1 [hunk 19]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+        - Configure the following Authentication password policy on each NSX manager.
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37447]:
```suggestion
        - Configure the following Authentication password policy on each NSX Manager node.
```

File: PowerValidatedSolutions.psm1 [hunk 20]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+		.SYNOPSIS
+        Retrieve the current Authentication Policy from NSX Manager Nodes
+
+        .DESCRIPTION
+        The Get-NsxtManagerAuthenticationPolicy cmdlet retrieves the current Authentication policy from each NSX
+        manager nodes for a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and
+        -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Manager
+        - Output the following Authentication policy on each NSX manager node.
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37394]:
```suggestion
        - Output the following Authentication policy on each NSX Local Manager node.
```

File: PowerValidatedSolutions.psm1 [hunk 21]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+			a) api_failed_auth_lockout_period (in sec)
+			b) api_failed_auth_reset_period (in sec)
+			c) api_max_auth_failures (in attempt)
+			d) cli_failed_auth_lockout_period (in sec)
+			e) cli_max_auth_failures (in attempt)
+			f) minimum_password_length (in characters)
+
+        .EXAMPLE
+        Set-NsxtManagerAuthenticationPolicy -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -apiLockoutPeriod 900 -apiResetPeriod 120 -apiMaxAttempt 5 -cliLockoutPeriod 900 -cliMaxAttempt 5 -minPasswdLength 15
+        This example configures the Authentication password policy in NSX manager nodes in sfo-m01 workload domain
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37457]:
```suggestion
        This example configures the Authentication password policy in NSX Local Manager nodes in sfo-m01 workload domain
```

File: PowerValidatedSolutions.psm1 [hunk 22]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+    <#
+		.SYNOPSIS
+        Retrieve Authentication Policy from NSX Edge Nodes
+
+        .DESCRIPTION
+        The Get-NsxtEdgeNodeAuthenticationPolicy cmdlet retrieves the current Authentication policy from NSX Edge
+        nodes within a workload domain. The cmdlet connects to SDDC Manager using the -server, -user, and -password
+        values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Management Cluster
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37535]:
```suggestion
        - Validates that network connectivity and authentication is possible to NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 23]
```diff
@@ -37078,5 +37380,271 @@ Function Get-SsoPasswordPolicies {
...
+Function Set-NsxtEdgeNodeAuthenticationPolicy {
+    <#
+		.SYNOPSIS
+        Configure Authentication Policy NSX Edge Nodes
+
+        .DESCRIPTION
+        The Set-NsxtEdgeNodeAuthenticationPolicy cmdlet configures the Authentication policy within NSX Edge nodes.
+        The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to NSX Management Cluster
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:37586]:
```suggestion
        - Validates that network connectivity and authentication is possible to NSX Local Manager
```

File: PowerValidatedSolutions.psm1 [hunk 24]
```diff
@@ -17087,6 +17145,79 @@ Function Request-NsxtEdgeAccountLockout {
...
+                                                    $updatedConfiguration = Get-LocalPasswordComplexity -vmName $nsxtEdgeNode.display_name -guestUser $vcfNsxDetails.rootUser -guestPassword $vcfNsxDetails.rootPass -nsx
+                                                    if ($updatedConfiguration.'Min Length' -eq $minLength -and $updatedConfiguration.'Min Lowercase' -eq $minLowercase -and $updatedConfiguration.'Min Uppercase' -eq $minUppercase -and $updatedConfiguration.'Min Numerical' -eq $minNumerical -and $updatedConfiguration.'Min Special' -eq $minSpecial -and $updatedConfiguration.'Min Unique' -eq $minUnique -and $updatedConfiguration.'Max Retries' -eq $maxRetry) {
+                                                        Write-Output "Update Password Complexity Policy on NSX Manager Node ($($nsxtEdgeNode.display_name)): SUCCESSFUL"
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17196]: Edge Node?

```suggestion
                                                        Write-Output "Update Password Complexity Policy on NSX Edge Node ($($nsxtEdgeNode.display_name)): SUCCESSFUL"
```

**Review by tenthirtyam — COMMENTED** [review]
:octocat: I'll defer the merge to you post-commits so that you can do so on your time zone. 

Nice work! 🚀

**Pull request merged**